### PR TITLE
Fix regex `get_data_files` formatting for base paths

### PR DESCRIPTION
**Pull request by ZachNagengast**
With this pr https://github.com/huggingface/datasets/pull/6309, it is formatting the entire base path into regex, which results in the undesired formatting error `doesn't match the pattern` because of the line in `glob_pattern_to_regex`: `.replace("//", "/")`:
- Input: `hf://datasets/...`
- Output: `hf:/datasets/...`

This fix will only convert the `split_pattern` to regex and keep the `base_path` unchanged.

cc @albertvillanova hopefully this still works with your implementation

### CODE REVIEW

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks for the proposed fix, @ZachNagengast.

The reason why I used the the `glob_pattern_to_regex` in the entire `pattern` is because otherwise I got an error for Windows local paths: a `base_path` like `'C:\\Users\\runneradmin...` makes the function `string_to_dict` raise `re.error: incomplete escape \U at position 2`

See: https://github.com/huggingface/datasets/actions/runs/6544904352/job/17772361643

We should include a test that includes the case you mention, and find a solution that works for all cases.

**Comment by ZachNagengast** [comment]
> The reason why I used the the glob_pattern_to_regex in the entire pattern is because otherwise I got an error for Windows local paths: a base_path like 'C:\\Users\\runneradmin... made the function string_to_dict raise re.error: incomplete escape \U at position 2

What is the expected inputs and outputs for the windows `base_path`

**Comment by lhoestq** [comment]
We took the liberty of continuing this PR to include it it today's patch release :)
I hope you don't mind

File: src/datasets/data_files.py [hunk 1]
```diff
@@ -244,8 +244,10 @@ def _get_data_files_patterns(
         except FileNotFoundError:
             continue
         if len(data_files) > 0:
-            pattern = base_path + ("/" if base_path else "") + split_pattern
-            splits: Set[str] = {string_to_dict(p, glob_pattern_to_regex(pattern))["split"] for p in data_files}
+            splits: Set[str] = {
+                string_to_dict(xbasename(p), glob_pattern_to_regex(xbasename(split_pattern)))["split"]
+                for p in data_files
+            }
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/data_files.py:250]: This is for old `push_to_hub` to work: they push custom splits using this pattern in the `data` directory.
New `push_to_hub` have some YAML to specify the pattern to use, so `get_data_patterns` isn't called

[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/data_files.py:250]: OK, all clear now. Thanks.

**Review by albertvillanova — APPROVED** [review]

**Pull request merged**